lms/djangoapps/teams/csv.py

```diff
@@ -19,3 +27,199 @@ def load_team_membership_csv(course, response):
+                    continue
+                user = self.get_user(username)
+                if user is None:
+                    continue
+                if not self.validate_user_enrolled_in_course(user):
+                    row['user'] = None
+                    continue
+                row['user'] = user
+
+                if not self.validate_user_assignment_to_team_and_teamset(row):
+                    return False
+                row_dictionaries.append(row)
+
+            if not self.validation_errors:
+                for row in row_dictionaries:
+                    self.add_user_to_team(row)
+                return True
+            else:
+                return False
+        return False
+
+    def validate_teamsets(self):
+        """
+        Validates team set names. Returns true if there are no errors.
+        The following conditions result in errors:
+        Teamset does not exist
+        Teamset id is duplicated
+        Also populates the teamset_names_list.
+        header_row is the list of teamset_ids
+        """
+        teamset_ids = {ts.teamset_id for ts in self.course_module.teams_configuration.teamsets}
+        dupe_set = set()
+        for teamset_id in self.teamset_ids:
+            if teamset_id in dupe_set:
+                self.validation_errors.append("Teamset with id " + teamset_id + " is duplicated.")
+                return False
+            dupe_set.add(teamset_id)
+            if teamset_id not in teamset_ids:
+                self.validation_errors.append("Teamset with id " + teamset_id + " does not exist.")
+                return False
+            self.user_ids_by_teamset_id[teamset_id] = {m.user_id for m in CourseTeamMembership.objects.filter
+                                                       (team__course_id=self.course.id, team__topic_id=teamset_id)}
+        return True
+
+    def validate_user_enrolled_in_course(self, user):
+        """
+        Invalid states:
+            user not enrolled in course
+        """
+        if not CourseEnrollment.is_enrolled(user, self.course.id):
+            self.validation_errors.append('User ' + user.username + ' is not enrolled in this course.')
+            return False
+
+        return True
+
+    def validate_user_assignment_to_team_and_teamset(self, row):
+        """
+        Validates a user entry relative to an existing team.
+        row is a dictionary where key is column name and value is the row value
+        [andrew],masters,team1,,team3
+        [joe],masters,,team2,team3
+        """
+        user = row['user']
+        for teamset_id in self.teamset_ids:
+            team_name = row[teamset_id]
+            if not team_name:
+                continue
+            try:
+                # checks for a team inside a specific team set. This way team names can be duplicated across
+                # teamsets
+                team = CourseTeam.objects.get(name=team_name, topic_id=teamset_id, course_id=self.course.id)
```

Comment:
We're going to end up making this query (number of rows) * (number of teamsets) times.  It might be worth making one query for every `CourseTeam` in this course after `validate_teamsets()` returns True, then creating a mapping from (team_name, topic_id) -> CourseTeam.

Reply:
to make this more visible, django `TestCase` provides self.assertNumQueries, you should toss one of those in one of your tests